src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+}
+
+void ui_tile::draw( WINDOW *win, int x, int y ) const
+{
+    mvwputch( win, x, y, color, sym );
+}
+
+ui_tile::ui_tile( long sym, nc_color color ) : sym( sym ), color( color )
+{
+}
+
+tabbed_window::tabbed_window(size_t size_x, size_t size_y, int x, int y, ui_anchor anchor) : bordered_window(size_x, size_y, x, y, anchor), tab_index(0)
+{
+}
+
+void tabbed_window::local_draw()
+{
+    bordered_window::local_draw();
+    auto win = get_win(); // never null
+    //erase the top 3 rows
+    for( unsigned int y = 0; y < 3; y++ ) {
+        for( unsigned int x = 0; x < get_rect().size_x; x++ ){
+            mvwputch( win, y, x, border_color, ' ' );
+        }
+    }
+
+    for ( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+        mvwputch( win, 2, i, border_color, LINE_OXOX );
+    }
+
+    mvwputch( win, 2,  0, border_color, LINE_OXXO ); // |^
+    mvwputch( win, 2, get_rect().size_x - 1, border_color, LINE_OOXX ); // ^|
+
+    int x_offset = 1; // leave space for selection bracket
+    for( unsigned int i = 0; i < tabs.size(); i++ ) {
+        x_offset += draw_tab( win, x_offset, tabs[i].first, tab_index == i ) + 2;
+    }
+}
+
+template<class T>
+T *tabbed_window::create_tab( std::string tab )
+{
+    T *tab_win = new T( get_rect().size_x - 2, get_rect().size_y - 4, 1, -1, bottom_left );
+    tabs.push_back( {tab, tab_win} );
+    tab_win->set_visible( tabs.size() == 1 ? true : false );
+    add_child( tab_win );
+    return tab_win;
+}
+
+void tabbed_window::next_tab()
+{
+    tabs[tab_index].second->set_visible( false );
+    if( tab_index == tabs.size() - 1 ){
+        tab_index = 0;
+    } else {
+        tab_index++;
+    }
+    tabs[tab_index].second->set_visible( true );
+}
+
+void tabbed_window::previous_tab()
+{
+    tabs[tab_index].second->set_visible( false );
+    if( tab_index == 0 ){
+        tab_index = tabs.size() - 1;
+    } else {
+        tab_index--;
+    }
+    tabs[tab_index].second->set_visible( true );
+}
+
+const std::pair<std::string, ui_window *> &tabbed_window::current_tab() const
+{
+    return tabs[tab_index];
+}
+
+auto_bordered_window::auto_bordered_window( size_t size_x, size_t size_y, int x , int y, ui_anchor anchor ) : ui_window( size_x, size_y, x, y, anchor )
+{
+    uncovered = new bool[size_x * size_y];
+    recalc_uncovered();
+}
+
+auto_bordered_window::~auto_bordered_window()
+{
+    delete[] uncovered;
+}
+
+void auto_bordered_window::set_rect( const ui_rect &new_rect )
+{
+    ui_window::set_rect( new_rect );
+    delete[] uncovered;
+    uncovered = new bool[new_rect.size_x * new_rect.size_y];
+    recalc_uncovered();
+}
+
+void auto_bordered_window::add_child( ui_element *child )
+{
+    ui_window::add_child( child );
+    recalc_uncovered();
+}
+
+void auto_bordered_window::recalc_uncovered()
+{
+    for( unsigned int x = 0; x < get_rect().size_x; x++ ) {
+        for( unsigned int y = 0; y < get_rect().size_y; y++ ) {
+            uncovered[y * get_rect().size_x + x] = true;
```

Comment:
It might be useful to crate a simple array_2d wrapper class for this. It's used here for the bool vector and above for the tiles, and it may be used by other code as well.